### PR TITLE
Add IRPGO and CSIRPGO options to Swift

### PR DESCRIPTION
**Pull request by chiragramani**
This PR introduces three new instrumentation flags and plumbs them through to IRGen:

1. `-ir-profile-generate` - enable IR-level instrumentation.
2. `-cs-profile-generate` - enable context-sensitive IR-level instrumentation.
3. `-ir-profile-use` - IR-level PGO input profdata file to enable profile-guided optimization (both IRPGO and CSIRPGO)

**Context:** https://forums.swift.org/t/ir-level-pgo-instrumentation-in-swift/82123

**Swift-driver PR:** https://github.com/swiftlang/swift-driver/pull/1992

**Tests and validation:**
This PR includes ir level verification tests, also checks few edge-cases when `-ir-profile-use` supplied profile is either missing or is an invalid IR profile. 

However, for argument validation, linking, and generating IR profiles that can later be consumed by -cs-profile-generate, I’ll need corresponding swift-driver changes. Those changes are being tracked in https://github.com/swiftlang/swift-driver/pull/1992

The plan is:

1. Land this PR first.
2. Then update swift-driver with the required changes.
4. Then add the remaining set of test cases in this repo.

~~**Open Questions (Reviewer Input Requested)**~~

~~1. **Naming:**~~
~~Is `-ir-profile-generate` a good flag name?
`-cs-profile-generate` pairs nicely with Clang’s `-fcs-profile-generate`. However, `-ir-profile-generate` feels closer to Clang's `-fprofile-generate` but `-profile-generate` already exists in Swift as a frontend level and not IR level. Should it stay as is, or is there a clearer alternative?~~

~~2. **Missing LLVM runtime + Blocked on tests**~~

~~At the moment, I haven’t added tests because they currently fail locally due to missing LLVM runtime symbols. This is expected, since the required runtime linking through libclang APIs will be handled in swift-driver(https://github.com/swiftlang/swift-driver/pull/1992). However, swift-driver itself depends on these new options being available in Swift.~~

~~Does that mean the correct order of the PRs should be the following?~~

~~1. Create a PR in Swift that introduces just the new options, and land that PR first.
2. Land the corresponding swift-driver PR to handle runtime linking,
3. Then follow up with a Swift PR adding the full functionality and tests?~~
~~

### CODE REVIEW

**Comment by aschwaighofer** [comment]
I don't have a strong opinion on naming. My feeling is that context sensitive and non context sensitive LLVM IR level instrumentation should follow a similar pattern to make clear they are related. We have already deviated from clang's naming scheme -- that is, clang's `fprofile-generate`(llvm ir level instrumentation) is not equal swift's `-profile-generate` ("AST" based insertion of instrumentation). `-ir-profile-generate` and `-cs-ir-profile-generate` would make it clearer that they are related.

**Comment by aschwaighofer** [comment]
> Does that mean the correct order of the PRs should be the following?
>  Create a PR in Swift that introduces just the new options, and land that PR first.
>  Land the corresponding swift-driver PR to handle runtime linking,
>  Then follow up with a Swift PR adding the full functionality and tests?

You could add LLVM IR level tests that test the functionality of this PR (look at most tests in `test/IRGen`) and merge this PR first. Then you could land the corresponding swift-driver PR. Then you could at a `swiftlang/swift` repo commit to add execution tests. 

```
// RUN: %target-swift-frontend -ir-profile-generate -emit-ir %s | %FileCheck %s

// CHECK: define @mangleNameOfA
// CHECK: the llvm ir you expect to see (calls to profile runtime function/intrinsics)
public func a(_ cond: Bool) {
  if cond {
     b()
  } else {
    c()
  }
}
```

Preferably, you would as `.sil` tests rather than `.swift` tests as that would make them more "stable".

That would look something like the following (I ran `swiftc -emit-sil test.swift` and edited the output a little bit):

```
// RUN: %target-swift-frontend -ir-profile-generate -emit-ir %s | %FileCheck %s

sil_stage canonical

import Builtin
import Swift
import SwiftShims

sil @b : $@convention(thin) () -> ()

sil @c : $@convention(thin) () -> ()

// CHECK: define {{.*}} void @a
// CHECK: the llvm ir you expect to see (calls to profile runtime function/intrinsics)

sil @a : $@convention(thin) (Bool) -> () {
bb0(%0 : $Bool):
  %2 = struct_extract %0, #Bool._value
  cond_br %2, bb1, bb2

bb1:
  // function_ref b()
  %4 = function_ref @b : $@convention(thin) () -> ()
  %5 = apply %4() : $@convention(thin) () -> ()
  br bb3 

bb2:
  // function_ref c()
  %7 = function_ref @c: $@convention(thin) () -> ()
  %8 = apply %7() : $@convention(thin) () -> ()
  br bb3 

bb3:
  %10 = tuple ()
  return %10
}
```

File: lib/IRGen/IRGen.cpp [hunk 1]
```diff
@@ -231,6 +231,37 @@ static void populatePGOOptions(std::optional<PGOOptions> &Out,
...
+    Out = PGOOptions(
+      /*ProfileFile=*/ hasUse ? Opts.UseProfile : "",
+      /*CSProfileGenFile=*/ Opts.CSProfileGenFile,
+      /*ProfileRemappingFile=*/ "",
+      /*MemoryProfile=*/ "",
+      /*FS=*/ llvm::vfs::getRealFileSystem(),
+      /*Action=*/ hasUse ? PGOOptions::IRUse : PGOOptions::NoAction,
+      /*CSPGOAction=*/ PGOOptions::CSIRInstr,
+      /*ColdType=*/ PGOOptions::ColdFuncOpt::Default,
+      /*DebugInfoForProfiling=*/ Opts.DebugInfoForProfiling
```
Comment (ellishg) [inline lib/IRGen/IRGen.cpp]: NIT: I don't think we need these comments if the argument is obvious from the name.

```suggestion
      /*ProfileFile=*/ hasUse ? Opts.UseProfile : "",
      Opts.CSProfileGenFile,
      /*ProfileRemappingFile=*/ "",
      /*MemoryProfile=*/ "",
      /*FS=*/ llvm::vfs::getRealFileSystem(),
      hasUse ? PGOOptions::IRUse : PGOOptions::NoAction,
      PGOOptions::CSIRInstr,
      PGOOptions::ColdFuncOpt::Default,
     Opts.DebugInfoForProfiling
```

File: lib/IRGen/IRGen.cpp [hunk 2]
```diff
@@ -231,6 +231,37 @@ static void populatePGOOptions(std::optional<PGOOptions> &Out,
     return;
   }
 
+  if (Opts.EnableCSIRProfileGen) {
+    const bool hasUse = !Opts.UseProfile.empty();
+    Out = PGOOptions(
+      /*ProfileFile=*/ hasUse ? Opts.UseProfile : "",
```
Comment (ellishg) [inline lib/IRGen/IRGen.cpp]: The check isn't necessary.

```suggestion
      /*ProfileFile=*/ Opts.UseProfile,
```

File: lib/IRGen/IRGen.cpp [hunk 3]
```diff
@@ -231,6 +231,37 @@ static void populatePGOOptions(std::optional<PGOOptions> &Out,
     return;
   }
 
+  if (Opts.EnableCSIRProfileGen) {
+    const bool hasUse = !Opts.UseProfile.empty();
```
Comment (ellishg) [inline lib/IRGen/IRGen.cpp]: Should we create a new `IRPGOProfileGenFile` variable instead of reusing `UseProfile`? `UseProfile` is used for Swift frontend-instrumentation and will be a different profile than LLVM's IRPGO.

File: lib/IRGen/IRGen.cpp [hunk 4]
```diff
@@ -231,6 +231,37 @@ static void populatePGOOptions(std::optional<PGOOptions> &Out,
...
+      /*FS=*/ llvm::vfs::getRealFileSystem(),
+      /*Action=*/ hasUse ? PGOOptions::IRUse : PGOOptions::NoAction,
+      /*CSPGOAction=*/ PGOOptions::CSIRInstr,
+      /*ColdType=*/ PGOOptions::ColdFuncOpt::Default,
+      /*DebugInfoForProfiling=*/ Opts.DebugInfoForProfiling
+    );
+    return;
+  }
+
+  if (Opts.EnableIRProfileGen) {
```
Comment (ellishg) [inline lib/IRGen/IRGen.cpp:299]: Can you add cases for `EnableIRProfileUse` and `EnableCSIRProfileUse`?

Comment (chiragramani) [reply lib/IRGen/IRGen.cpp:299]: Yes, my initial thought was to handle the IR-only usage modes in a follow-up PR, but I’m happy to include them here if that’s preferable. I was leaning that way since Clang uses `IndexedInstrProfReader` to decide between IR and CSIR profiles and doesn't have a dedicated input flag which takes in intent ir/cs usage. 

Would you recommend taking a similar approach in Swift as part of this PR?

[clang/lib/Frontend/CompilerInvocation.cpp#L1497](https://github.com/llvm/llvm-project/blob/44b7abcc75b005ab87e11e2beac155bf0b155992/clang/lib/Frontend/CompilerInvocation.cpp#L1497)

**Comment by chiragramani** [comment]
> Preferably, you would as `.sil` tests rather than `.swift` tests as that would make them more "stable".

> You could add LLVM IR level tests that test the functionality of this PR (look at most tests in `test/IRGen`) and merge this PR first. Then you could land the corresponding swift-driver PR. Then you could at a `swiftlang/swift` repo commit to add execution tests.

Thanks for the guidance here, it definitely helped me understand things better!

File: lib/IRGen/IRGen.cpp (hunk 4 above)
Comment (chiragramani) [reply lib/IRGen/IRGen.cpp:299]: Added those changes in this PR so it’s easier for everyone to review in one place.
Please let me know if I’ve missed anything :)

Comment (ellishg) [reply lib/IRGen/IRGen.cpp:299]: Actually sorry I forgot there is no -fcs-profile-use. We can leave -profile-use to a separate PR if you want.

Comment (chiragramani) [reply lib/IRGen/IRGen.cpp:299]: No worries at all and thanks for pointing it out, makes sense to include `-ir-profile-use` in this PR. 👍

**Comment by chiragramani** [comment]
Hi @aschwaighofer, I've added the tests based on your suggestions. When you have a moment, could you please review the changes and share your feedback? Thanks

File: include/swift/Option/Options.td [hunk 5]
```diff
@@ -1608,6 +1608,24 @@ def profile_sample_use : Joined<["-"], "profile-sample-use=">,
...
+  HelpText<"Generate instrumented code to collect context sensitive execution counts into default.profraw (overridden by LLVM_PROFILE_FILE env var)">;
+
+def cs_profile_generate_EQ : Joined<["-"], "cs-profile-generate=">,
+  Flags<[FrontendOption, NoInteractiveOption]>,
+  MetaVarName<"<directory>">,
+  HelpText<"Generate instrumented code to collect context sensitive execution counts into <directory>/default.profraw (overridden by LLVM_PROFILE_FILE env var)">;
+
+def ir_profile_generate: Flag<["-"], "ir-profile-generate">,
+  Flags<[FrontendOption, NoInteractiveOption]>,
+  HelpText<"Generate instrumented code to collect execution counts into default.profraw (overridden by LLVM_PROFILE_FILE env var)">;
```
Comment (ellishg) [inline include/swift/Option/Options.td:1622]: Can we also add `-ir-profile-generate=`?

File: test/IRGen/cs_profile_generate.sil [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+
+// CHECK: br i1 {{.*}}, label %[[THEN:[0-9]+]], label %[[ELSE:[0-9]+]]
+// THEN: increment counter[0] and call b()
+// CHECK: [[THEN]]:
+// CHECK:   %[[L0:.*]] = load i64, ptr @__profc_a, align 8
+// CHECK:   %[[A0:.*]] = add i64 %[[L0]], 1
+// CHECK:   store i64 %[[A0]], ptr @__profc_a, align 8
+// CHECK:   call swiftcc void @b()
+// CHECK:   br label %[[MERGE:[0-9]+]]
```
Comment (ellishg) [inline test/IRGen/cs_profile_generate.sil:36]: I'm not sure we need to check how exactly instrumentation is lowered. It should be enough to check that we run the `PGOInstrumentationGen` pass. Can we do something like that here?

https://github.com/llvm/llvm-project/blob/8fcb712167fa809af9bec2ffcfe1ac1b8c5e4aa6/clang/test/CodeGen/pgo-instrumentation.c#L3-L5

File: include/swift/Option/Options.td (hunk 5 above)
Comment (chiragramani) [reply include/swift/Option/Options.td:1622]: I forgot about that one :)
Added, thanks for pointing it out!

File: test/IRGen/cs_profile_generate.sil (hunk 6 above)
Comment (chiragramani) [reply test/IRGen/cs_profile_generate.sil:36]: For sure, yes, that would be a good check to have. 
Added, thank you!

**Comment by chiragramani** [comment]
@kavon would love to hear your perspective on this PR as well, when you have a moment.

**Comment by chiragramani** [comment]
@swift-ci test

**Comment by chiragramani** [comment]
Hey @aschwaighofer, would it be possible for you to tag or add the respective POCs who could help review this PR please? I’d really appreciate it.

**Comment by drodriguez** [comment]
@swift-ci please test

**Comment by chiragramani** [comment]
@swift-ci please test

**Comment by drodriguez** [comment]
@swift-ci please test

**Comment by drodriguez** [comment]
> please test

@chiragramani sadly you need a special role for starting the tests. The details are in https://www.swift.org/contributing/#member (still some hoops to jump through, but easier than before)

**Comment by aschwaighofer** [comment]
@swift-ci test

**Review by aschwaighofer — COMMENTED** [review]
LGTM

**Review by aschwaighofer — APPROVED** [review]

**Comment by drodriguez** [comment]
@swift-ci please test

**Comment by chiragramani** [comment]
> @chiragramani sadly you need a special role for starting the tests. The details are in https://www.swift.org/contributing/#member (still some hoops to jump through, but easier than before)

Really appreciate you helping trigger the tests and sharing the process. I’ll send in a request soon.

**Comment by aschwaighofer** [comment]
@swift-ci please test

**Comment by chiragramani** [comment]
> [Swift Test Linux Platform](https://ci.swift.org/job/swift-PR-Linux/24265/) — Failed

I noticed that the above job has failed, but I’m unable to open the link for the Linux job details. Could someone with access kindly share the logs or error message? Thanks

**Comment by drodriguez** [comment]
ci.swift.org seems to give me time-outs. It might be having a bad day.

**Comment by chiragramani** [comment]
> ci.swift.org seems to give me time-outs. It might be having a bad day.

Thanks for checking. It looks like the link works now, but the error seems related to a CI infra issue (agent removal). Would it be possible to retry the job?

**Comment by drodriguez** [comment]
@swift-ci please test Linux platform

**Comment by chiragramani** [comment]
> @swift-ci please test Linux platform

Thanks! The smoke test ran post this but the Swift Test Linux Platform didn’t get picked up. Can we give it another run/retry please  :pray:?

**Comment by drodriguez** [comment]
The "(smoke test)" gets fulfilled by the non smoke test. It seems, however, that your comment quote did trigger the Linux Platform build in https://ci.swift.org/job/swift-PR-Linux/24357/ so I cannot check the previous results, but hopefully those will turn green.

(I have a couple of PRs and I am also in testing hell, but in my case with macOS failing randomly all the time… patience)

**Comment by chiragramani** [comment]
```
******************** TEST 'Swift(linux-x86_64) :: Interop/Cxx/class/access/swiftify-private-fileid.swift' FAILED ********************
15:30:00  couldn't parse text: '://162106643'
15:30:00  in expression: 'rdar://162106643'
15:30:00  in REQUIRES: directive on test line 1
15:30:00  ********************
```
It failed :sweat_smile:
I saw that the failure is in swiftify-private-fileid.swift, but it looks unrelated to this PR changes. Any suggestions how to go about it?

**Comment by drodriguez** [comment]
#84775 should take care of those. It has been one of those weeks.

**Comment by drodriguez** [comment]
@swift-ci please test Linux platform

**Comment by chiragramani** [comment]
Thank you for all your help @drodriguez , its all green now :raised_hands:

File: include/swift/Option/Options.td [hunk 7]
```diff
@@ -1608,6 +1608,29 @@ def profile_sample_use : Joined<["-"], "profile-sample-use=">,
...
+
+def ir_profile_generate_EQ : Joined<["-"], "ir-profile-generate=">,
+  Flags<[FrontendOption, NoInteractiveOption]>,
+  MetaVarName<"<directory>">,
+  HelpText<"Generate instrumented code to collect execution counts into <directory>/default.profraw (overridden by LLVM_PROFILE_FILE env var)">;
+
+def ir_profile_use : CommaJoined<["-"], "ir-profile-use=">,
```
Comment (cachemeifyoucan) [inline include/swift/Option/Options.td:1629]: This file needs to be captured by `SwiftDependencyTracker` so it is available when swift caching is enabled.
